server/src/main/java/io/druid/metadata/SQLMetadataSupervisorManager.java

```diff
@@ -0,0 +1,250 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.metadata;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Supplier;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.inject.Inject;
+import com.metamx.common.Pair;
+import com.metamx.common.lifecycle.LifecycleStart;
+import io.druid.guice.ManageLifecycle;
+import io.druid.guice.annotations.Json;
+import io.druid.indexing.overlord.supervisor.SupervisorSpec;
+import io.druid.indexing.overlord.supervisor.VersionedSupervisorSpec;
+import org.joda.time.DateTime;
+import org.skife.jdbi.v2.FoldController;
+import org.skife.jdbi.v2.Folder3;
+import org.skife.jdbi.v2.Handle;
+import org.skife.jdbi.v2.IDBI;
+import org.skife.jdbi.v2.StatementContext;
+import org.skife.jdbi.v2.tweak.HandleCallback;
+import org.skife.jdbi.v2.tweak.ResultSetMapper;
+
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Map;
+
+@ManageLifecycle
+public class SQLMetadataSupervisorManager implements MetadataSupervisorManager
+{
+  private final ObjectMapper jsonMapper;
+  private final SQLMetadataConnector connector;
+  private final Supplier<MetadataStorageTablesConfig> dbTables;
+  private final IDBI dbi;
+
+  @Inject
+  public SQLMetadataSupervisorManager(
+      @Json ObjectMapper jsonMapper,
+      SQLMetadataConnector connector,
+      Supplier<MetadataStorageTablesConfig> dbTables
+  )
+  {
+    this.jsonMapper = jsonMapper;
+    this.connector = connector;
+    this.dbTables = dbTables;
+    this.dbi = connector.getDBI();
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    connector.createSupervisorsTable();
+  }
+
+  @Override
+  public void insert(final String id, final SupervisorSpec spec)
+  {
+    dbi.withHandle(
+        new HandleCallback<Void>()
+        {
+          @Override
+          public Void withHandle(Handle handle) throws Exception
+          {
+            final String version = new DateTime().toString();
```

Comment:
We don't limit history for any other kind of metadata (rules, segments, tasks) so I don't think we should limit history here. users could delete old rows if they really want to, or it might also make sense to introduce a retention option for more kinds of metadata in a future PR.